### PR TITLE
Switch to AntD TreeTable and remove deprecated code

### LINKED ISSUE #27: Consider Ant Design Table as UI Framework
Type: Feature Request

In [PR #24](https://github.com/eclipse-cdt-cloud/vscode-peripheral-inspector/pull/24), we are introducing a replacement for the Microsoft Tree View. This new implementation aims to closely align with the existing Tree API while extending it to support tree tables, facilitating easier migration and offering enhanced control and capabilities. The UI rendering is handled by [PrimeReact](https://primereact.org/), focusing primarily on the [Tree](https://primereact.org/tree/) and [TreeTable](https://primereact.org/treetable/) components.

During the PR review, we identified potential performance issues with the TreeTable component, particularly due to the lack of [Virtual Scrolling](https://github.com/primefaces/primereact/issues/1423). As a result, we opted to use the Tree component instead, which offers better performance. However, this choice introduces some functional and visual limitations, particularly the loss of the column-like behavior inherent to TreeTable.

To address this, I propose investigating alternatives to PrimeReact's TreeTable, specifically considering the [Ant Design Table](https://ant.design/components/table/) as a potential replacement. Since other functionalities, such as [search](https://github.com/eclipse-cdt-cloud/vscode-peripheral-inspector/issues/23), [inline-editing](https://github.com/eclipse-cdt-cloud/vscode-peripheral-inspector/issues/16), and [custom controls](https://github.com/eclipse-cdt-cloud/vscode-peripheral-inspector/issues/22), are partially dependent on the UI framework, it would be prudent to conduct this investigation before further feature development.

I would like to open this topic for discussion.

### PR DESCRIPTION
**Pull request by haydar-metin**
Closes #27

https://github.com/user-attachments/assets/c5d2998b-86f3-4596-9673-7e724b7bb1f9

## Tree

This PR removes the PrimeReact `Tree` and `TreeTable`, as well as the VSCode Tree. Instead, it introduces the Ant Design (AntD) Tree Table, which is allows virtualization and provides a simpler API. Alongside this, all deprecated code has been removed, ensuring the project now exclusively uses the new component.

## Model

One major issue with the slow rendering was the back-and-forth communication between the VSCode extension and the webview. To address this, the peripheral model structure has been separated into three layers:

- **Peripheral Model** (e.g., `PeripheralNode`): Handles data loading and updating.
- **Serialization Model** (e.g., `PeripheralNodeDTO`): Contains the serialized data of the peripheral model, which is transferred to the webview.
- **Display Model** (e.g., `CDTTreeItem`): Renders the tree table in the webview.

UI-specific aspects have been moved to a custom converter (e.g., `PeripheralNodeConverter`). This converter uses the serialized data to create the tree for rendering (**Serialization Model** -> **Display Model**).

The webview only sees the **Serialization Model / Display Model**.

### Communication

UI-specific actions like expansion and pinning are handled directly in the webview, which updates the display model immediately. However, these updates are also propagated back to the extension to update the domain model when needed. 

The webview can request a resync if necessary. In such cases, after an action is triggered, the entire serialized domain model is sent back to the webview to recreate the display model. 

This means expansion happens instantly, and if any data changes, the extension can sync the local model accordingly.

## Follow-up

1) I’ve tried to be as cautious as possible and avoided changing the peripheral model too much. It still includes "pinned" and "expansion" properties for now, but we should consider removing these in a follow-up PR.

2) While the tree feels smoother now, there’s still room for improvement. For instance, caching the peripheral models and serializing only the changes could make it even faster.

3) Instead of saving the expanded/pinned state directly in the **Peripheral Model**. We should move it to a custom service.

4) Most of the commands are not necessary anymore (e.g., `Pin`, `Set Format`). We should remove them and handle those cases directly in the webview. For example, we could show a context menu with the different formats instead of using the VSCode quick select. 

PS: Most of the changes are moving functionality around.

### CODE REVIEW

File: src/common/format.ts [hunk 1]
```diff
@@ -24,3 +24,42 @@ export function formatValue(value: number, hexLength: number, format: NumberForm
             return hexFormat(value, hexLength, true);
     }
 }
+
+export function toStringDecHexOctBin(val: number/* should be an integer*/): string {
```
Comment (jreineckearm) [inline src/common/format.ts:28]: This came in from a file with a different copyright notice.

File: src/common/peripheral.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License as outlined in the LICENSE File
+ ********************************************************************************/
+
+/**
+ * Those peripheral names will be ignored in the peripheral tree.
+ * They will not be resolved from the SVD file.
+ */
+export const IGNORE_PERIPHERAL_NAMES = ['AXBS0', 'CAU3'];
```
Comment (jreineckearm) [inline src/common/peripheral.ts:12]: I start liking the idea of having an ignore list. But we probably shouldn't hard-code it. Rather make it configurable.

File: src/common/vscode.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License as outlined in the LICENSE File
+ ********************************************************************************/
+import { NumberFormat } from './format';
+
+export interface VscodeContext {
+    'data-vscode-context': string;
+}
+
+export interface NodeSetting {
```
Comment (jreineckearm) [inline src/common/vscode.ts:13]: Coming from a file with a different copyright notice.

File: src/manifest.ts [hunk 4]
```diff
@@ -15,7 +15,7 @@ export const DEFAULT_DEVICE = 'deviceName';
 export const CONFIG_PROCESSOR = 'processorConfig';
 export const DEFAULT_PROCESSOR = 'processorName';
 export const CONFIG_ADDRGAP = 'svdAddrGapThreshold';
-export const DEFAULT_ADDRGAP = 16;
+export const DEFAULT_ADDRGAP = -1;
```
Comment (jreineckearm) [inline src/manifest.ts:18]: Setting to `-1` is useful for testing purposes as discussed "offline". But we should not change the default on `main`.

File: package.json [hunk 5]
```diff
@@ -305,7 +307,7 @@
         },
         "peripheral-inspector.svdAddrGapThreshold": {
           "type": "number",
-          "default": 16,
+          "default": -1,
```
Comment (jreineckearm) [inline package.json:310]: Setting to `-1` is useful for testing purposes as discussed "offline". But we should not change the default on `main`.

File: src/plugin/peripheral/tree/peripheral-session-tree.ts [hunk 6]
```diff
@@ -150,9 +149,9 @@ export class PeripheralTreeForSession extends PeripheralBaseNodeImpl {
             const provider = this.api.getPeripheralsProvider(svdPath);
             if (provider) {
                 const enumTypeValuesMap = {};
-                const poptions = await provider.getPeripherals(data, { gapThreshold });
-                peripherials = poptions.map((options) => new PeripheralNodeImpl(gapThreshold, options));
-                peripherials.sort(PeripheralNode.compare);
+                const poptions = (await provider.getPeripherals(data, { gapThreshold })).filter(p => !IGNORE_PERIPHERAL_NAMES.includes(p.name));
```
Comment (jreineckearm) [inline src/plugin/peripheral/tree/peripheral-session-tree.ts:152]: See comment at `IGNORE_PERIPHERAL_NAMES` definition.

File: src/svd-parser.ts [hunk 7]
```diff
@@ -83,7 +83,9 @@ export class SVDParser {
 
         svdData.device.peripherals[0].peripheral.forEach((element) => {
             const name = element.name[0];
-            peripheralMap[name] = element;
+            if (!IGNORE_PERIPHERAL_NAMES.includes(name)) {
```
Comment (jreineckearm) [inline src/svd-parser.ts:86]: See comment at `IGNORE_PERIPHERAL_NAMES` definition.

**Review by jreineckearm — COMMENTED** [review]
See some comments.

**Comment by jreineckearm** [comment]
One finding from testing:
* I see this behavior if I reduce the width of the panel so that there is not enough space to show all text of the name column:
image

**Comment by jreineckearm** [comment]
Some more findings:
* With the original TreeView implementation it was possible to navigate via the keyboard. And hence to select a node. We should try to restore that. This may also help with navigating the tree in combination with the search functionality.
* We have now two additional context menu entries `Undo` and `Redo` (compared to TreeView implementation). What history would these use? Only user input? Or also values from updates due to executing code? We might want to consider removing those for the time being. If they haven't been added by purpose.
* The display of pinned values looks slightly different (compared to the TreeView and the PrimeReact implementation). Is it possible to remove the extra node by adding the pin icon inline?
Before
image
After
image

File: src/common/peripheral.ts (hunk 2 above)
Comment (martin-fleck-at) [reply src/common/peripheral.ts:12]: How would you like this to be configurable? A VS Code setting or should that be something on the API level?

**Comment by haydar-metin** [comment]
Thanks for the feedback @jreineckearm!

The latest changes introduced:
- ellipsis for the columns and other css fixes
- settings for ignored peripheral names
- keyboard navigation (same way as in vscode)

**Comment by jreineckearm** [comment]
Thanks, @haydar-metin !

* The rendering artifacts caused by the panel resizing are gone. I noticed however that narrowing the panel size comes with a delay after releasing the mouse button before the view is re-rendered. Whereas widening the panel size takes immediate effect. Is this something under our control?

* The keyboard navigation works nicely. However, I believe the associated node selection causes some trouble. To reproduce, click into the tree, then click on the vertical scrollbar and drag up and down. At some point the scrollbar keeps moving if I release the left mouse button. It requires another click into the tree to stop that.
I noticed this rendering effect which makes me wonder if there is an issue with the mouse-hit-areas of vertical scrollbar and node elements due to some miscalculated boundaries....this one unfortunately needs attention before merge.
image

* The new `undo` /`redo` context menu entries are still there. But I could live with that.

* The pinning works like before the refurbishing. Thanks for that!

**Comment by martin-fleck-at** [comment]
> * The rendering artifacts caused by the panel resizing are gone. I noticed however that narrowing the panel size comes with a delay after releasing the mouse button before the view is re-rendered. Whereas widening the panel size takes immediate effect. Is this something under our control?

Very interesting, not sure what could cause this. However, I also experience this with other views in VS Code (e.g., Copilot Chat) so I'm not sure what we can do about this.

> * The keyboard navigation works nicely. However, I believe the associated node selection causes some trouble. To reproduce, click into the tree, then click on the vertical scrollbar and drag up and down. At some point the scrollbar keeps moving if I release the left mouse button. It requires another click into the tree to stop that.
>   I noticed this rendering effect which makes me wonder if there is an issue with the mouse-hit-areas of vertical scrollbar and node elements due to some miscalculated boundaries....this one unfortunately needs attention before merge.
>   image

I pushed a commit that should fix that. From what I can tell, the problem is that when you release the mouse outside the iframe (our view), we do not abort the scrolling as we do not receive the event. I now abort the scrolling on purpose once we leave the iframe, could you please check whether that also fixes your problem?

> * The new `undo` /`redo` context menu entries are still there. But I could live with that.

Haydar and I had a look at this together but we couldn't find any way to suppress this. There is some code in that allows you to suppress [cut/copy/paste](https://github.com/microsoft/vscode/blob/ae176274ac92eb4c0bfb93343eca18e4ebb46ea7/src/vs/workbench/contrib/webview/browser/webview.contribution.ts#L49) but not undo/redo. I can also see the Undo/Redo in other extensions webview, e.g., GitLens, so we probably need to add a hook in VS Code for that.

**Comment by martin-fleck-at** [comment]
@jreineckearm I think that sensitivity might have always been there. Basically we do not receive any events from outside the iframe. I pushed a commit that aborts the scroll on re-enter which gives a much better user experience in my opinion - the scrolling still stops outside the view but if you keep the mouse pressed and go back it will continue. Could you have a look whether that works for you as well?

**Comment by jreineckearm** [comment]
Thanks, @martin-fleck-at ! Much better!
I compared also with the behavior of other views in VS Code and can see that those even jump back to the start of the scrolling action. But I am good with not replicating that behavior right now. Could be a future enhancement (unless you say is super-easy to do this :-) ).

File: package.json [hunk 8]
```diff
@@ -315,6 +245,13 @@
           "type": "boolean",
           "default": true,
           "description": "Save layout of peripheral view between sessions"
+        },
+        "peripheral-inspector.ignorePeripherals": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "List of case insensitive peripherals names to ignore"
```
Comment (jreineckearm) [inline package.json:254]:
```suggestion
          "description": "List of case insensitive peripheral names to ignore"
```

**Comment by jreineckearm** [comment]
The peripheral ignore list works nicely. Some future enhancements around this:
* Allow modification of the list at runtime.
* Allow adding peripherals to the list through Peripheral Inspector context menu (maybe there for workspace not user settings).
  * Clearing the list might be worth a consideration to keep symmetry in the feature.

**Comment by martin-fleck-at** [comment]
@jreineckearm Sounds good to me, should we create a separate issue for the enhancements? I pushed your typo fix as well. If you agree with the PR now, please feel free to approve and merge it ;-)

**Review by jreineckearm — APPROVED** [review]
Thanks again for this! Changes look good to me now. Nice work!

@martin-fleck-at , yes, please open the remaining enhancement ideas as new issues.

@planger , @thegecko , could you please help with merging this PR?

Thanks,
Jens

**Comment by planger** [comment]
@jreineckearm :rocket: Great! Happy to merge. I guess we want to squash, right?

**Comment by jreineckearm** [comment]
Yes, squash it was :-)

**Pull request merged**